From f064314b950888c4ba2c4803aecbf531ccc16dd1 Mon Sep 17 00:00:00 2001
From: "Philip, Avinash" <avinashphilip@ti.com>
Date: Mon, 27 Feb 2012 11:30:40 +0530
Subject: [PATCH 102/172] ARM: OMAP2+: gpmc: Add low power support

This patch adds Callback functions to support low power transitions by
adding
1. pm_runtime API's to handle GPMC power transitions
2. GPMC save/restore context.

Signed-off-by: Philip, Avinash <avinashphilip@ti.com>
---
 arch/arm/mach-omap2/gpmc-nand.c        |    2 ++
 arch/arm/mach-omap2/gpmc.c             |   18 ++++++++++++++++++
 arch/arm/plat-omap/include/plat/gpmc.h |    2 ++
 arch/arm/plat-omap/include/plat/nand.h |    2 ++
 4 files changed, 24 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-omap2/gpmc-nand.c b/arch/arm/mach-omap2/gpmc-nand.c
index c12b286..0207a66 100644
--- a/arch/arm/mach-omap2/gpmc-nand.c
+++ b/arch/arm/mach-omap2/gpmc-nand.c
@@ -108,6 +108,8 @@ int __devinit gpmc_nand_init(struct omap_nand_platform_data *gpmc_nand_data)
 	}
 
 	gpmc_nand_device.dev.platform_data = gpmc_nand_data;
+	gpmc_nand_data->ctrlr_suspend	= gpmc_suspend;
+	gpmc_nand_data->ctrlr_resume	= gpmc_resume;
 
 	printk(KERN_INFO "Registering NAND on CS%d\n", gpmc_nand_data->cs);
 
diff --git a/arch/arm/mach-omap2/gpmc.c b/arch/arm/mach-omap2/gpmc.c
index 50940c0..c58a336 100644
--- a/arch/arm/mach-omap2/gpmc.c
+++ b/arch/arm/mach-omap2/gpmc.c
@@ -740,6 +740,8 @@ static void __devinit gpmc_mem_init(void)
 	}
 }
 
+struct device *gpmc_dev;
+
 static int __devinit gpmc_probe(struct platform_device *pdev)
 {
 	u32 l;
@@ -751,6 +753,8 @@ static int __devinit gpmc_probe(struct platform_device *pdev)
 	/* XXX: This should go away with HWMOD & runtime PM adaptation */
 	gpmc_clk_init(&pdev->dev);
 
+	gpmc_dev = &pdev->dev;
+
 	gpmc = devm_kzalloc(&pdev->dev, sizeof(struct gpmc), GFP_KERNEL);
 	if (!gpmc)
 		return -ENOMEM;
@@ -825,6 +829,20 @@ static struct platform_driver gpmc_driver = {
 
 module_platform_driver(gpmc_driver);
 
+int gpmc_suspend(void)
+{
+	omap3_gpmc_save_context();
+	pm_runtime_put_sync(gpmc_dev);
+	return 0;
+}
+
+int gpmc_resume(void)
+{
+	pm_runtime_get_sync(gpmc_dev);
+	omap3_gpmc_restore_context();
+	return 0;
+}
+
 #ifdef CONFIG_ARCH_OMAP3
 static struct omap3_gpmc_regs gpmc_context;
 
diff --git a/arch/arm/plat-omap/include/plat/gpmc.h b/arch/arm/plat-omap/include/plat/gpmc.h
index 506b3a1..2b81fc6 100644
--- a/arch/arm/plat-omap/include/plat/gpmc.h
+++ b/arch/arm/plat-omap/include/plat/gpmc.h
@@ -174,4 +174,6 @@ extern int gpmc_nand_write(int cs, int cmd, int wval);
 
 int gpmc_enable_hwecc(int ecc, int cs, int mode, int dev_width, int ecc_size);
 int gpmc_calculate_ecc(int ecc, int cs, const u_char *dat, u_char *ecc_code);
+int gpmc_suspend(void);
+int gpmc_resume(void);
 #endif
diff --git a/arch/arm/plat-omap/include/plat/nand.h b/arch/arm/plat-omap/include/plat/nand.h
index 1bcff1d..9212a2e 100644
--- a/arch/arm/plat-omap/include/plat/nand.h
+++ b/arch/arm/plat-omap/include/plat/nand.h
@@ -30,6 +30,8 @@ struct omap_nand_platform_data {
 	int			devsize;
 	enum omap_ecc           ecc_opt;
 	bool			elm_used;
+	int (*ctrlr_suspend) (void);
+	int (*ctrlr_resume) (void);
 };
 
 /* minimum size for IO mapping */
-- 
1.7.9.1

